jerry-ext/module/module.c

```diff
@@ -0,0 +1,215 @@
+{
+  jerry_value_t ret = jerry_create_error (error_type, message);
+  jerry_value_t property_name = jerry_create_string (module_name_property_name);
+  jerry_value_t property_value = jerry_create_string_from_utf8 (module_name);
+
+  jerry_release_value (jerry_set_property (ret, property_name, property_value));
+  jerry_release_value (property_name);
+  jerry_release_value (property_value);
+  return ret;
+} /* jerryx_module_create_error */
+
+
+/**
+ * Initialize the module manager extension.
+ */
+static void
+jerryx_module_manager_init (void *user_data_p)
+{
+  jerry_value_t context = jerry_create_object ();
+  *((jerry_value_t *) user_data_p) = jerry_acquire_value (context);
+  jerry_release_value (context);
+} /* jerryx_module_manager_init */
+
+/**
+ * Deinitialize the module manager extension.
+ */
+static void
+jerryx_module_manager_deinit (void *user_data_p) /**< context pointer to deinitialize */
+{
+  jerry_release_value (*(jerry_value_t *) user_data_p);
+} /* jerryx_module_manager_deinit */
+
+/**
+ * Declare the context data manager for modules.
+ */
+static const jerry_context_data_manager_t jerryx_module_manager =
+{
+  .init_cb = jerryx_module_manager_init,
+  .deinit_cb = jerryx_module_manager_deinit,
+  .bytes_needed = sizeof (jerry_value_t)
+};
+
+/**
+ * Declare the linker section where module definitions are stored.
+ */
+JERRYX_SECTION_DECLARE (jxm_modules, jerryx_module_t)
+
+/**
+ * Attempt to retrieve a module by name from a cache, and return zero if not found.
+ */
+static jerry_value_t
+jerryx_module_check_cache (jerry_value_t cache, /**< cache from which to attempt to retrieve the module by name */
+                          jerry_value_t module_name) /* JerryScript string value holding the module name */
+{
+  /* Check if the cache has the module. */
+  jerry_value_t ret = jerry_has_property (cache, module_name);
+
+  /* If we succeed in getting an answer, we examine the answer. */
+  if (!jerry_value_has_error_flag (ret))
+  {
+    bool has_property = jerry_get_boolean_value (ret);
+
+    /* Clean up the JerryScript value holding the boolean. */
+    jerry_release_value (ret);
+    ret = 0;
+
+    /* If the module is indeed in the cache, we return it. */
+    if (has_property)
+    {
+      ret = jerry_get_property (cache, module_name);
+    }
+  }
+
+  return ret;
+} /* jerryx_module_check_cache */
+
+/**
+ * Attempt to cache a loaded module.
+ *
+ * @return the module on success, otherwise the error encountered when attempting to cache. In the latter case, the
+ * @p module is released.
+ */
+static jerry_value_t
+jerryx_module_add_to_cache (jerry_value_t cache, /**< cache to which to add the module */
+                            jerry_value_t module_name, /**< key at which to cache the module */
+                            jerry_value_t module) /**< the module to cache */
+{
+  jerry_value_t ret = jerry_set_property (cache, module_name, module);
```

Comment:
Using a jerry object avoids having to use `jmem_heap_alloc_block ()`. I can certainly revert to using `jmem_heap_alloc_block ()` but what data structure should I use, especially if we're considering that there may be many modules? I was using a linked list before, but that would probably be even more inefficient.

I'm certainly happy to implement, like a prefix tree or something.